### PR TITLE
sequence_task_1 dspektor

### PR DESCRIPTION
**Pull request by spektorius**
image

### CODE REVIEW

File: sequence_task_1/student/include/impl.h [hunk 1]
```diff
@@ -20,31 +21,84 @@ class StaticPlaylist
...
+    using size_type       = typename Container::size_type;
+    using allocator_type  = typename Container::allocator_type;
+    using difference_type = typename Container::difference_type;
+    using iterator        = typename Container::iterator;
+    using const_iterator  = typename Container::const_iterator;
+    using reference       = typename Container::reference;
+    using const_reference = typename Container::const_reference;
 
     /** @todo Iterators */
+    const_iterator begin()
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:34]: Mark as const

File: sequence_task_1/student/include/impl.h [hunk 2]
```diff
@@ -20,31 +21,84 @@ class StaticPlaylist
...
+        m_tracklist.assign(otherConatiner.cbegin(), otherConatiner.cend());
+        return m_tracklist;
+    }
 
     /** @todo Add track from initializer */
     template<class... Args>
-    const Song& play(Args&&... songData);
+    const Song& play(Args&&... songData)
+    {
+        return *m_tracklist.insert(end(), {Song(std::forward<Args>(songData))...});
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:62]: You don't need to construct a temporary object, if you already have the initialization data, that can be used for in-place construction.

File: sequence_task_1/student/include/impl.h [hunk 3]
```diff
@@ -20,31 +21,84 @@ class StaticPlaylist
...
 
     /** @todo Skip to the next track in playlist, remove current */
-    void switchNext();
+    void switchNext()
+    {
+        if (!hasTracks())
+        {
+            throw std::out_of_range("Playlist is empty");
+        }
+        m_tracklist.erase(begin());
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:88]: **Not effective for vector**

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: sequence_task_1/student/include/impl.h (hunk 3 above)
Comment (spektorius) [reply sequence_task_1/student/include/impl.h]: fixed by reversing container on creation or assignment

File: sequence_task_1/student/include/impl.h [hunk 4]
```diff
@@ -31,41 +31,41 @@ class StaticPlaylist
...
+    const_iterator end() const
     {
         return m_tracklist.cend();
     }
 
     /** @todo Constructor from any reversible sequence container */
     template<class OtherContainer>
     StaticPlaylist(const OtherContainer& otherConatiner)
-        : m_tracklist(otherConatiner.cbegin(), otherConatiner.cend())
+        : m_tracklist(otherConatiner.crbegin(), otherConatiner.crend())
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:47]: Yeah, right!

File: sequence_task_1/student/include/impl.h [hunk 5]
```diff
@@ -31,41 +31,41 @@ class StaticPlaylist
...
+        m_tracklist.assign(otherConatiner.crbegin(), otherConatiner.crend());
         return m_tracklist;
     }
 
     /** @todo Add track from initializer */
     template<class... Args>
     const Song& play(Args&&... songData)
     {
-        return *m_tracklist.insert(end(), {Song(std::forward<Args>(songData))...});
+        return *m_tracklist.emplace(begin(), Song(std::forward<Args>(songData))...);
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:62]: Still temporary 'Song' object :(
also, why emplace is at begin? still not effective for vector

File: sequence_task_1/student/include/impl.h [hunk 6]
```diff
@@ -31,41 +31,41 @@ class StaticPlaylist
...
     }
 
     /** @todo Add track */
     const Song& play(const Song& song)
     {
-        return m_tracklist.push_back(song);
+        return *m_tracklist.emplace(begin(), song);
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:68]: Not effective for vector

File: sequence_task_1/student/include/impl.h (hunk 5 above)
Comment (spektorius) [reply sequence_task_1/student/include/impl.h:62]: Not clear, as we have to insert tracks to the other end of the tracks queue.
Also, the Song constructor is marked as explicit so I need to call it anyway.

Comment (smay1613) [reply sequence_task_1/student/include/impl.h:62]: Please verify with https://github.com/smay1613/LTC-STL-Student/pull/111/files

**Pull request closed without merging**